ext/mbstring/mbstring.c

```diff
@@ -1316,95 +1316,70 @@ PHP_FUNCTION(mb_internal_encoding)
 /* {{{ Returns the input encoding */
 PHP_FUNCTION(mb_http_input)
 {
-	char *typ = NULL;
-	size_t typ_len = 0;
-	int retname = 1;
-	char *list, *temp;
-	const mbfl_encoding *result = NULL;
+	char *type = NULL;
+	size_t type_len = 0, n;
+	const mbfl_encoding **entry;
 
 	ZEND_PARSE_PARAMETERS_START(0, 1)
 		Z_PARAM_OPTIONAL
-		Z_PARAM_STRING(typ, typ_len)
+		Z_PARAM_STRING(type, type_len)
 	ZEND_PARSE_PARAMETERS_END();
 
-	if (typ == NULL) {
-		result = MBSTRG(http_input_identify);
+	if (type == NULL) {
+		RETVAL_STRING(MBSTRG(http_input_identify)->name);
 	} else {
-		switch (*typ) {
+		switch (*type) {
 		case 'G':
 		case 'g':
-			result = MBSTRG(http_input_identify_get);
+			RETVAL_STRING(MBSTRG(http_input_identify_get)->name);
 			break;
 		case 'P':
 		case 'p':
-			result = MBSTRG(http_input_identify_post);
+			RETVAL_STRING(MBSTRG(http_input_identify_post)->name);
 			break;
 		case 'C':
 		case 'c':
-			result = MBSTRG(http_input_identify_cookie);
+			RETVAL_STRING(MBSTRG(http_input_identify_cookie)->name);
 			break;
 		case 'S':
 		case 's':
-			result = MBSTRG(http_input_identify_string);
+			RETVAL_STRING(MBSTRG(http_input_identify_string)->name);
 			break;
 		case 'I':
 		case 'i':
-			{
-				const mbfl_encoding **entry = MBSTRG(http_input_list);
-				const size_t n = MBSTRG(http_input_list_size);
-				size_t i;
-				array_init(return_value);
-				for (i = 0; i < n; i++) {
-					add_next_index_string(return_value, (*entry)->name);
-					entry++;
-				}
-				retname = 0;
+			entry = MBSTRG(http_input_list);
+			n = MBSTRG(http_input_list_size);
+			array_init(return_value);
+			while (n--) {
+				add_next_index_string(return_value, (*entry++)->name);
```

Comment:
I think we should avoid these kinds of "clever" loops. A normal incrementing induction variable is easier to understand. Also not a fan of inline increments if we can avoid them.